lib/ads/ad_utils.js

```diff
@@ -7,15 +7,134 @@
 
 goog.provide('shaka.ads.Utils');
 
+goog.require('shaka.util.TextParser');
+goog.require('shaka.util.TXml');
+
 
 /**
  * A class responsible for ad utils.
  * @export
  */
 shaka.ads.Utils = class {
-
+  /**
+   * @param {!shaka.extern.xml.Node} vast
+   * @param {?number} currentTime
+   * @return {!Array.<shaka.extern.AdInterstitial>}
+   */
+  static parseVastToInterstitials(vast, currentTime) {
+    const TXml = shaka.util.TXml;
+    /** @type {!Array.<shaka.extern.AdInterstitial>} */
+    const interstitials = [];
+
+    let startTime = 0;
+    if (currentTime != null) {
+      startTime = currentTime;
+    }
+
+    for (const ad of TXml.findChildren(vast, 'Ad')) {
+      const inline = TXml.findChild(ad, 'InLine');
+      if (!inline) {
+        continue;
+      }
+      const creatives = TXml.findChild(inline, 'Creatives');
+      if (!creatives) {
+        continue;
+      }
+      for (const creative of TXml.findChildren(creatives, 'Creative')) {
+        const linear = TXml.findChild(creative, 'Linear');
+        if (!linear) {
+          continue;
+        }
+        let skipOffset = null;
+        if (linear.attributes['skipoffset']) {
+          skipOffset = shaka.ads.Utils.parseTime(
+              linear.attributes['skipoffset']);
+          if (isNaN(skipOffset)) {
+            skipOffset = null;
+          }
+        }
+        const mediaFiles = TXml.findChild(linear, 'MediaFiles');
+        if (!mediaFiles) {
+          continue;
+        }
+        const medias = TXml.findChildren(mediaFiles, 'MediaFile');
+        let checkMedias = medias;
+        const streamingMedias = medias.filter((media) => {
+          return media.attributes['delivery'] == 'streaming';
+        });
+        if (streamingMedias.length) {
+          checkMedias = streamingMedias;
+        }
+        const sortedMedias = checkMedias.sort((a, b) => {
+          const aHeight = parseInt(a.attributes['height'], 10) || 0;
+          const bHeight = parseInt(b.attributes['height'], 10) || 0;
+          return bHeight - aHeight;
+        });
+        for (const media of sortedMedias) {
+          const adUrl = TXml.getTextContents(media);
+          if (!adUrl) {
+            continue;
+          }
+          interstitials.push({
+            id: null,
+            startTime: startTime,
+            endTime: null,
+            uri: adUrl,
+            isSkippable: skipOffset != null,
+            skipOffset,
+            canJump: false,
+            resumeOffset: 0,
+            playoutLimit: null,
+            once: true,
+            pre: currentTime == null,
+            post: currentTime == Infinity,
+            timelineRange: false,
+          });
+          break;
+        }
+      }
+    }
+    return interstitials;
+  }
+
+  /**
+   * Parses a time from string.
+   *
+   * @param {?string} time
+   * @return {?number}
+   */
+  static parseTime(time) {
```

Comment:
Could you make this a method on `shaka.util.TextParser`, and then use it in `shaka.text.SbvTextParser` and `shaka.text.Cue`?

Those two classes both have their own timestamp parsers that work mostly the same as this one, they just don't accept timestamps that lack a milliseconds field.

Reply:
Done!